pkg/model/multiversion/delta.go

```diff
@@ -0,0 +1,307 @@
+	FieldChangeTypeNone                           FieldChangeType = "none"
+	FieldChangeTypeAdded                          FieldChangeType = "added"
+	FieldChangeTypeRemoved                        FieldChangeType = "removed"
+	FieldChangeTypeRenamed                        FieldChangeType = "renamed"
+	FieldChangeTypeShapeChanged                   FieldChangeType = "shape-changed"
+	FieldChangeTypeShapeChangedFromStringToSecret FieldChangeType = "shape-changed-from-string-to-secret"
+	FieldChangeTypeShapeChangedFromSecretToString FieldChangeType = "shape-changed-from-secret-to-string"
+)
+
+// FieldDelta represent the delta between the same original field in two
+// different CRD versions. If a field is removed in the Destination version
+// the Destination value will be nil. If a field is new in the Destination
+// version, the Source value will be nil.
+type FieldDelta struct {
+	ChangeType FieldChangeType
+	// Field from the destination CRD
+	Destination *ackmodel.Field
+	// Field from the source CRD
+	Source *ackmodel.Field
+}
+
+// CRDDelta stores the spec and status deltas for a custom resource.
+type CRDDelta struct {
+	SpecDeltas   []FieldDelta
+	StatusDeltas []FieldDelta
+}
+
+// ComputeCRDFieldsDeltas compares two ackmodel.CRD instances and returns the
+// spec and status fields deltas. srcCRD is the CRD of the spoke (source) version
+// and dstCRD is the CRD of the hub (destination) version.
+func ComputeCRDFieldsDeltas(srcCRD, dstCRD *ackmodel.CRD) (*CRDDelta, error) {
+	dstRenames, err := dstCRD.GetAllRenames(ackmodel.OpTypeCreate)
+	if err != nil {
+		return nil, fmt.Errorf("cannot get resource field renames: %s", err)
+	}
+
+	srcRenames, err := srcCRD.GetAllRenames(ackmodel.OpTypeCreate)
+	if err != nil {
+		return nil, fmt.Errorf("cannot get resource field renames: %s", err)
+	}
+
+	renames, err := ComputeRenamesDelta(srcRenames, dstRenames)
+	if err != nil {
+		return nil, fmt.Errorf("cannot compute the field renames delta: %v", err)
+	}
+
+	specDeltas, err := ComputeFieldsDelta(srcCRD.SpecFields, dstCRD.SpecFields, renames)
+	if err != nil {
+		return nil, fmt.Errorf("cannot compute spec fields deltas: %s", err)
+	}
+
+	statusDeltas, err := ComputeFieldsDelta(srcCRD.StatusFields, dstCRD.StatusFields, renames)
+	if err != nil {
+		return nil, fmt.Errorf("cannot compute status fields deltas: %s", err)
+	}
+
+	return &CRDDelta{
+		specDeltas,
+		statusDeltas,
+	}, nil
+}
+
+// fieldChangedToSecret returns true if field changed from string to secret.
+func fieldChangedToSecret(src, dst *ackmodel.Field) bool {
+	return (src.FieldConfig == nil ||
+		(src.FieldConfig != nil && !src.FieldConfig.IsSecret)) &&
+		(dst.FieldConfig != nil && dst.FieldConfig.IsSecret)
+}
+
+// ComputeFieldsDelta computes the difference between two maps of fields. It returns a list
+// of FieldDelta's that contains the ChangeType and at least one field reference.
+func ComputeFieldsDelta(
```

Comment:
A couple more naming changes please:

1. Please name this function `ComputeFieldDeltas` not `ComputeFieldsDelta`
2. Please change `srcCRDFields` to `srcFields` and `dstCRDFields` to `dstFields`. A Field is part of a CRD so there's no reason to repeat `CRD` in the variable name.

Reply:
Changed